File: benchmarks/fibonacci/dynamic.jl

include("../../timing/dphpc_timing.jl")

function fib(n)
    if n <= 2
        return 1
    end
    sₙ = 1
    sₙ₋₁ = 1
    for i=3:n
        sₙ₋₁, sₙ = sₙ, sₙ + sₙ₋₁ 
    end
    return sₙ
end

function main()

    # nothing as first argument because there is no reset/init code needed here
    # but still need to pass a first argument

    # these are the two ways to call macros in julia
    @dphpc_time(nothing, fib(1000),  "S")  # with brackets, like a function 
    @dphpc_time nothing  fib(10000)  "M"   # or just with some space between the arguments
    @dphpc_time nothing  fib(100000) "L"   # whithout the brackets, everything needs to be on the same line

    # there is also a 2 argument version like in C
    @dphpc_time(nothing, fib(38))
end



main()

File: benchmarks/fibonacci/dynamic.c
#include "../../timing/dphpc_timing.h"



long fib(int n) {
    if (n <= 2) {
        return 1;
    }
    long sn = 1;
    long sn_1 = 1;
    for (int i = 3; i <= n; i++) {
        long tmp = sn_1;
        sn_1 = sn;
        sn = sn + tmp;
    }
    return sn;
}

int main() {

    dphpc_time3(
        , // no reset/init code in this case
        fib(1000),
        "S"
    );

    dphpc_time3(
        ,
        fib(10000),
        "M"
    );

    dphpc_time3(
        ,
        fib(100000),
        "L"
    );

    dphpc_time3(
        ,
        fib(100500),
        "paper"
    );

    // preset, if given, has to be one of these 4: "S", "M", "L", "paper"
    // versions where you use dphpc_time or dphpc_time2 (without specifying the preset) are always run
}